File: php-linter.rb
# 複数のLinterでチェック

require 'parallel'
require 'open3'

PHPMD_DEFAULT = 'text codesize,design,unusedcode'

def main(argv)
  cmd = argv.shift
  case cmd
  when 'php-l'
    exec "#{__dir__}/internal/php-lint-run"
  when 'phan'
    exec "#{__dir__}/internal/phan-run"
  when 'phpmd'
    argv = argv.empty? ? PHPMD_DEFAULT : argv.join(' ')
    exec "#{__dir__}/internal/phpmd-run #{argv}"
  when 'multi'
    linterMulti argv.join(' ')
  when 'help'
    usage
  when /.+/
    usage
  else
    exec "#{__dir__}/internal/php-lint-run"
  end
end

def usage
  puts <<-EOS
usage) #{File.basename(__FILE__)} [<mode> | help]

mode:
  php-l   linter by php-lint [default]
  phan    linter by phan
  phpmd   linter by phpmd
  multi   multiple linter(phan, phpmd, and php-l)

help: this message
  EOS
end

def linterMulti(phpmdArgv)
  # php -l
  # NOTE: これは早いので先に済ませる
  lines = exec "#{__dir__}/internal/php-lint-run"
  if lines > 0
    exit 0
  end

  # NOTE: 以降は時間がかかるので並列実行
  
  argv = phpmdArgv.empty? ? PHPMD_DEFAULT : phpmdArgv.join(' ')
  # phpmd
  cmd_phpmd = "#{__dir__}/internal/phpmd-run #{argv}"
  # phan
  cmd_phan = "#{__dir__}/internal/phan-run"

  msgsLst = Parallel.map([cmd_phpmd, cmd_phan], in_threads: 2) do |cmd|
    `#{cmd}`
  end
  msgsLst.each do |_msgs|
    unless _msgs.empty?
      print _msgs
      exit 0
    end
  end
end

# @return 行数
def exec(cmd)
  lines = 0
  Open3.popen3(cmd) {|_, o|
    o.each_line do |line|
      puts line
      lines += 1
    end
  }
  lines
end

main ARGV

File: php-linter.py
# -*- coding: utf-8 -*-

import os
import sys
import subprocess
import re

from concurrent.futures import ThreadPoolExecutor
from threading import Lock

PHPMD_DEFAULT = 'text codesize,design,unusedcode'
cwd = os.path.dirname(__file__)


def main(argv):
    argv.pop(0)

    cmd = argv.pop(0) if len(argv) > 0 else ''
    cmds = cmd.split(',')

    for cmd in cmds:
        nline = 0
        if cmd == 'php-l':
            nline = exec(cwd+'/internal/php-lint-run')
        elif cmd == 'phan':
            nline = exec(cwd+'/internal/phan-run')
        elif cmd == 'phpmd':
            argv = PHPMD_DEFAULT if len(argv) == 0 else ' '.join(argv)
            nline = exec(cwd+'/internal/phpmd-run '+argv)
        elif cmd == 'multi':
            nline = linterMulti(argv)
        elif cmd == 'help':
            nline = usage()
        else:
            nline = 1
            print('unknown command:{}'.format(cmd))

        if nline > 0:
            break


def usage():
    str = '''
usage) {} [<mode> | help]

mode:
  php-l   linter by php-lint [default]
  phan    linter by phan
  phpmd   linter by phpmd
  multi   multiple linter(phan, phpmd, and php-l)

help: this message
    '''
    print(str.format(os.path.basename(__file__)))


def linterMulti(phpmdArgv):
    # php -l
    # NOTE: これは速いので先に済ませる
    nline = exec(cwd+'/internal/php-lint-run')
    if nline > 0:
        return nline

    # NOTE: 以降は時間がかかるので並列実行

    argv = PHPMD_DEFAULT if len(phpmdArgv) == 0 else ' '.join(phpmdArgv)
    # phpmd
    cmd_phpmd = cwd+'/internal/phpmd-run '+argv
    # phan
    cmd_phan = cwd+'/internal/phan-run'
    with ThreadPoolExecutor(max_workers=2) as executor:
        msgsList = executor.map(parallelCmd, [cmd_phpmd, cmd_phan])

    for msgs in msgsList:
        if len(msgs) > 0:
            for msg in msgs:
                print(msg)
            return len(msgs)
    return 0

def parallelCmd(cmd):
    lines = []

    proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
    while True:
        line = proc.stdout.readline()
        if not line:
            break
        lines.append(line.decode('utf-8').rstrip())

    return lines


# @return 行数
def exec(cmd):
    nline = 0
    proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
    buf = []

    while True:
        line = proc.stdout.readline()
        if not line:
            break

        print(line.decode('utf-8').rstrip())
        nline += 1

    return nline


if __name__ == '__main__':
    main(sys.argv[:])
